primitives/runtime/src/traits.rs

```diff
@@ -1210,18 +1216,23 @@ pub trait Header: Clone + Send + Sync + Codec + Eq + MaybeSerialize + Debug + 's
 	}
 }
 
+/// Something that provides the Header Type.
+pub trait HeaderProvider {
+	/// Header type.
+	type Header: Header<Hash = Self::Hash>;
+	/// Header hash type.
+	type Hash: HashOutput;
+}
```

Comment:
Wait, why do we need this?

Reply:
This is needed to fix the "cyclical" issue in loading Header/BlockNumber as part of a `pallet::call`. Essentially, `construct_runtime` aggregates all calls to create a `RuntimeCall` that is then used to define `UncheckedExtrinsic`. This `UncheckedExtrinsic` is supplied to the Block. So, if we do not create a trait outside of Block that doesn't have Extrinsic, we go into a recursive loop leading to a build error.

So, this trait extracts Header from the trait Block and is then used to define the following:
```rust
/// The Header type used by the runtime.
pub type HeaderOf<T> = <<T as Config>::Block as traits::HeaderProvider>::Header;
/// The block number type used by the runtime.
pub type BlockNumberOf<T> = <HeaderOf<T> as traits::Header>::Number;
```

Reply:
> So, if we do not create a trait outside of Block that doesn't have Extrinsic, we go into a recursive loop leading to a build error.

Can you please show me the build error? 

I tried compiling `f4e772a64bf4b0a2f050b4674178169254604a35` and it worked?

Reply:
the trait bound `sp_runtime::generic::UncheckedExtrinsic<MultiAddress<AccountId32, ()>, RuntimeCall, MultiSignature, (CheckNonZeroSender<Runtime>, CheckSpecVersion<Runtime>, CheckTxVersion<Runtime>, CheckGenesis<Runtime>, CheckEra<Runtime>, CheckNonce<Runtime>, CheckWeight<Runtime>, ChargeTransactionPayment<Runtime>)>: Encode` is not satisfied

Even though we do have Encode implemented for UncheckedExtrinsic
pub struct UncheckedExtrinsic<Address, Call, Signature, Extra>
| ^^^^ required by this bound in UncheckedExtrinsic
this pointer marks to the Call param there

This is the error. It only happens if you use `BlockNumberOf` or `HeaderOf` as an argument in `#[pallet::call]`.

Reply:
When I added it as a parameter in one of the calls of pallet-timestamp, I saw this:

error[E0275]: overflow evaluating the requirement `<mock::Test as frame_system::Config>::Block == sp_runtime::generic::Block<sp_runtime::generic::Header<_, BlakeTwo256>, sp_runtime::generic::UncheckedExtrinsic<u64, RuntimeCall, ..., ...>>` --> frame/timestamp/src/mock.rs:39:11 |

Reply:
I spend a bit of time trying to solve the issue here and failed, would love to see @bkchr give it a try. @gupnik is `f4e772a64bf4b0a2f050b4674178169254604a35` the right commit to try this?